### ISSUE TITLE
Serde Support (behind a feature flag)

### ISSUE THREAD

**Issue by sachaarbonel** [u1roh/kd-tree#10]
Hi, thanks for this great library. I successfully built my tree, and I would like to save it to disk using [bincode](https://docs.rs/bincode/latest/bincode/). However, I'm getting this error
```
the trait bound `KdSliceN<EmbeddedBook, UInt<UInt<UTerm, B1>, B0>>: Serialize` is not satisfied
the following other types implement trait `Serialize`:
  &'a T
  &'a mut T
  ()
  (T0, T1)
  (T0, T1, T2)
  (T0, T1, T2, T3)
  (T0, T1, T2, T3, T4)
  (T0, T1, T2, T3, T4, T5)
and 183 others
```
let me know if this feature is out of scope for this library or not

this is my code for reference
```rust
use serde::{Deserialize, Serialize};
use serde_big_array::BigArray;
use kd_tree::{KdPoint, KdSlice2};

#[derive(Debug, Serialize, Deserialize)]
pub struct EmbeddedBook {
    pub title: String,

    pub author: String,

    pub summary: String,

    #[serde(with = "BigArray")]
    pub embeddings: [f32; 384],
}

impl KdPoint for EmbeddedBook {
    type Scalar = f32;
    type Dim = typenum::U2; // 2 dimensional tree.
    fn at(&self, k: usize) -> f32 {
        self.embeddings[k]
    }
}

#[derive(Serialize, Deserialize)]
struct BookSearcher<'a> {
    tree: &'a KdSlice2<EmbeddedBook>,
}

impl<'a> BookSearcher<'a> {
    fn new(embeddedbooks: &'a mut Vec<EmbeddedBook>) -> Self {
        let kdtree = kd_tree::KdSlice2::sort_by(embeddedbooks, |item1, item2, k| {
            item1.embeddings[k]
                .partial_cmp(&item2.embeddings[k])
                .unwrap()
        });
        BookSearcher { tree: kdtree }
    }
}
```

**Comment by u1roh** [comment]
* kd-tree crate doesn't support `serde` currently.
* `KdSliceN` cannot support `serde` because it is unsized and it must be used as a reference.
* `KdTreeN` may support `serde` in the future.
* I recommend to save the `Vec<EmbeddedBook>` after it is sorted by `KdSlice2::sort`. Although you need to call `KdSlice2::sort` again when you load the data from the disk, `KdSliceN::sort` works faster for the array already sorted.
  * I guess `std::mem::transmute::<&[EmbeddedBook], &KdSlice2<EmbeddedBook>>(slice)` works when the `slice` is already sorted in kd-tree manner, but I don't recommend it.

By the way, your code is a little redundant.
Since your `EmbeddedBook` type implements `KdPoint`, you can use `KdSlice2::sort` instead of `KdSlice2::sort_by`, as follows:
```rust
impl<'a> BookSearcher<'a> {
    fn new(embeddedbooks: &'a mut Vec<EmbeddedBook>) -> Self {
        let kdtree = kd_tree::KdSlice2::sort(embeddedbooks);
        BookSearcher { tree: kdtree }
    }
}
```

**Comment by u1roh** [comment]
@sachaarbonel I released v0.5.0, which supports `serde` feature.
Can you check and try the new version? https://crates.io/crates/kd-tree
